File: src/commands/utility/roleinfo.ts
import { SlashCommandBuilder, EmbedBuilder, ChatInputCommandInteraction, Role } from 'discord.js';

export default {
  data: new SlashCommandBuilder()
    .setName('roleinfo')
    .setDescription('View details about a role')
    .addRoleOption(option =>
      option.setName('role').setDescription('Role to inspect').setRequired(true)
    ),
  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const role = interaction.options.getRole('role', true) as Role;

    const embed = new EmbedBuilder()
      .setTitle(role.name)
      .setColor(role.hexColor || '#5865f2')
      .addFields(
        { name: 'ID', value: role.id, inline: true },
        { name: 'Color', value: role.hexColor || 'None', inline: true },
        { name: 'Members', value: role.members.size.toString(), inline: true },
        { name: 'Mentionable', value: role.mentionable ? 'Yes' : 'No', inline: true },
        { name: 'Hoisted', value: role.hoist ? 'Yes' : 'No', inline: true },
        { name: 'Position', value: role.position.toString(), inline: true },
        { name: 'Created', value: `<t:${Math.floor(role.createdTimestamp / 1000)}:R>`, inline: true }
      );

    await interaction.reply({ embeds: [embed], ephemeral: true });
  }
};

File: src/bot/commands/utility/roleinfo.rs
use crate::bot::{Context, Error};
use poise::serenity_prelude as serenity;

/// View details about a role
#[poise::command(slash_command, ephemeral, guild_only)]
pub async fn roleinfo(
    ctx: Context<'_>,
    #[description = "Role to inspect"] role: serenity::Role,
) -> Result<(), Error> {
    let created_ts = role.id.created_at().unix_timestamp();
    let colour = role.colour;

    let embed = serenity::CreateEmbed::new()
        .title(&role.name)
        .colour(colour)
        .field("ID", role.id.to_string(), true)
        .field("Color", format!("#{:06X}", colour.0), true)
        .field(
            "Mentionable",
            if role.mentionable { "Yes" } else { "No" },
            true,
        )
        .field("Hoisted", if role.hoist { "Yes" } else { "No" }, true)
        .field("Position", role.position.to_string(), true)
        .field("Created", format!("<t:{created_ts}:R>"), true);

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}
